src/Microsoft.Azure.Batch.SoftwareEntitlement/StoredEntitlementHostVerifier.cs

```diff
@@ -0,0 +1,30 @@
+using System.Collections.Concurrent;
+
+namespace Microsoft.Azure.Batch.SoftwareEntitlement
+{
+    /// <summary>
+    /// An in-memory store of entitlement IDs and the corresponding host ID values
+    /// with which they have been used. For ensuring that every entitlement is only
+    /// used by a single host.
+    /// </summary>
+    public class StoredEntitlementHostVerifier : IHostVerifier
+    {
+        private readonly ConcurrentDictionary<string, string> _lookup = new ConcurrentDictionary<string, string>();
+
+        /// <summary>
+        /// Performs a check to verify that the entitlement was not previously
+        /// associated with a different host.
+        /// </summary>
+        /// <param name="entitlement">The entitlement ID</param>
+        /// <param name="hostId">The host ID</param>
+        /// <returns>
+        /// True if the entitlement is new or previously associated with the same host,
+        /// false if it was previously associated with a different host.
+        /// </returns>
+        public bool Verify(NodeEntitlements entitlement, string hostId)
+        {
+            var storedHostId = _lookup.GetOrAdd(entitlement.Identifier, hostId);
+            return storedHostId == hostId;
```

Comment:
Should we be case sensitive here? I don't think we mention one way or the other in the spec.

Reply:
Don't know. Sounds like we need to be a bit more specific in the spec. My inclination would be to leave it as case-sensitive for now until we're more certain what format it'll take.

Reply:
Agreed - easier to loosen something than tighten it later.